### PR TITLE
Dataset page update

### PR DESCRIPTION
**Pull request by joshc0044**

### CODE REVIEW

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    // print stats
+    // window.console.debug('raw key: ', rawKey);
+    // window.console.debug('sorted data: ', data_sorted);
+    // window.console.debug('q1: ', q1);
+    // window.console.debug('q3: ', q3);
+    // window.console.debug('interQuantileRange: ', interQuantileRange);
+    // window.console.debug('median: ', median);
+    // window.console.debug('min (whisker): ', min);
+    // window.console.debug('max (whisker): ', max);
+    // use minimum/max values in data as lower & upper bounds for whisker lines
```
Comment (weixuanfu) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: For whisker plot, maybe just simply use minData/maxData as lower/upper bounds. `    let min = q1 - (1.5 * interQuantileRange);
    let max = q3 + (1.5 * interQuantileRange);` may be not needed.

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    let interQuantileRange = q3 - q1;
+    let min = q1 - (1.5 * interQuantileRange);
+    let max = q3 + (1.5 * interQuantileRange);
+    // min = -1;
+    // median = 0;
+    // max = 1;
+    let minData = Math.min(...data_sorted);
+    let maxData = Math.max(...data_sorted);
+
+    // print stats
```
Comment (weixuanfu) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: Codes need some clean up before merging.

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    //window.console.log('median', median);
+
+    let q3 = d3.quantile(data_sorted, .75);
+    let interQuantileRange = q3 - q1;
+    let min = q1 - (1.5 * interQuantileRange);
+    let max = q3 + (1.5 * interQuantileRange);
+    // min = -1;
+    // median = 0;
+    // max = 1;
+    let minData = Math.min(...data_sorted);
```
Comment (weixuanfu) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: why not use `minData =  d3.quantile(data_sorted, 0);`? Will the value be not the same?

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    .append("svg")
+      .attr("width", width + margin.left + margin.right)
+      .attr("height", height + margin.top + margin.bottom)
+      .attr("stroke", "none")
+    .style("fill", "none")
+    .append("g")
+      .attr("transform",
+            "translate(" + margin.left + "," + margin.top + ")");
+    // get stats
+    let data_sorted = valByRowObj[rawKey].sort(d3.ascending);
```
Comment (weixuanfu) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: To my knowledge, d3.quantile depends on sorted data. Please make sure the data is properly sorted here.

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx (hunk 2 above)
Comment (joshc0044) [reply lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: I'll remove this and do a bit more cleanup as well

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx (hunk 4 above)
Comment (joshc0044) [reply lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: I believe that is the case as well. 

`valByRowObj` is a javascript object; it's the dataset with the column as a key and the value is an array with all of that columns values. The data is sorted using javascripts native array sort method with d3.ascending - https://github.com/d3/d3-array#ascending

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarChart/index.jsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+      // for every entry in testSet, map keys to color
+      let colorObjList = [];
+      testSet.forEach((depVal, i) => {
+        // use https://github.com/d3/d3-scale-chromatic#schemePaired for 12 colors
+        // to select unique color per class in dataset
+        let colorString;
+        if(i < 12) {
+          colorString = d3.schemePaired[i];
+        } else {
+          // if more than 12 classes in dataset use sequential color range
```
Comment (joshc0044) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarChart/index.jsx]: also note that if there are more than 12 unique classes then every unique class value color will be from a spectrum - I did this as a way to guarantee each class gets a distinct color (using a spectrum there are infinite colors but the contrast between entries can be small). 

I can also use distinct colors in a range like for the first twelve too

**Comment by hjwilli** [comment]
I just rebuilt (`docker-compose build`) from this branch and hard-refreshed the browser, but I'm not seeing the charts.  Maybe you forgot to include a dependency?

image

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx (hunk 1 above)
Comment (joshc0044) [reply lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx]: I made some test plots with plotly.js and have that on a separate branch - https://github.com/EpistasisLab/pennai/tree/plotly_test_for_dataset_page

I posted some screenshots on the slack but just to be thorough I'll upload a few here as well 

image

The plotly boxplot whisker lines should be configurable; there are three options for specifying a boxplot's boxpoints in plotly - https://plot.ly/javascript/reference/#box-boxpoints

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarCharts/index.jsx [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+      data={testData}
+      layout={plotLayout}
+      config={plotConfig}
+    />)
+  }
+
+  render() {
+
+    let plotlyChart = this.createPlotlyStackedChart();
+    plotlyChart.chartCutoff ? plotlyChart = (<p>Too many value to plot</p>) : null;
```
Comment (hjwilli) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarCharts/index.jsx]: small typo, should be "Too many values to plot"

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarCharts/barChart.test.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import BarCharts from './';
```
Comment (hjwilli) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarCharts/barChart.test.js]: May want to rename the "BarCharts" component to something along the lines of "StackedBarChart" for clarity

File: lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarChart/index.jsx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    };
+    let plotlyBarCharData = [];
+    data_sorted.forEach(val => {
+      classCountObj[val] = classCountObj[val] ? ++classCountObj[val] : 1;
+    })
+
+    // for every entry in depColSet, map keys to color
+    let colorObj = {};
+    let colorList = [];
+    let colorStringList = [
```
Comment (hjwilli) [inline lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarChart/index.jsx]: Identical values for the color list in `colorStringList` and the color-to-name mapping variable are defined in both `BarChart/index.jsx` and `BarCharts/index.jsx`.  This color scheme should probably be defined once elseware and referenced from these components.

It's likely we will want to reference this color scheme for plots on other pages later.

File: lab/webapp/package.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
   "description": "your friendly machine learning assistant",
   "scripts": {
     "build": "webpack",
-    "build-dev": "webpack --watch",
+    "build-dev": "webpack --watch --mode=development",
```
[the remark this answers is not in the document]
Comment (joshc0044) [reply lab/webapp/package.json]: I believe the only way to invoke/call build-dev is connecting directly into the running instance of pennai_lab_1 through `docker exec` and and then manually running the build-dev command; in a production environment this should normally not occur but I can remove/revert this change if deemed prudent

**Comment by joshc0044** [comment]
I fixed the typo for "too many values", renamed `BarCharts` to `StackedBarChart` and moved the color list & obj to a seperate static file. 

There was also a mistake for selecting alternative colors when were more than 12 classes - a fix for this is also included in this commit, I tested it with the kykopt dataset from pmlb dataset with 18 distinct classes and each bar has its own color

File: lab/webapp/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (hjwilli) [reply lab/webapp/package.json]: You are correct, this line is fine.

**Comment by joshc0044** [comment]
I fixed the problems with the boxplots & stacked bar charts

For point 2 in Weixuan's comment above, the max number of values the stacked bar chart can display is currently set to 40. This limit is not technically necessary as the chart should theoretically work with an arbitrary number of unique values for that given category. But, for the sake of readability, there is an upper limit. If an alternative chart is preferable to the message 'Too many values to plot' that should be feasible as well; create a pie chart or something else deemed appropriate. 

And for point 3 - apparently it is currently not possible to modify the styling of the content displayed when hovering over the box plot (https://community.plot.ly/t/box-plot-and-hover-info/10341/2). The content of the hover text is different depending on whether the boxplot is just showing the mean or the mean and the standard deviation (https://plot.ly/javascript/box-plots/#box-plot-styling-mean-and-standard-deviation)

**Comment by lacava** [comment]
- [ ] styling comments:
    - [ ] the text is too small and there is too much whitespace

- [ ] boxplots
    - [ ] the SD diamons are *kind of* cool, but probably just confusing since they aren't used regularly. I would just show the mean dotted line. 
    - [ ] the xticks are too small
    - [ ] i think it would look better with a border on the box plots
    - [ ] aspect ratio: the plots should be shorter and wider so that you can see more plots in browser at a time, much like openML
    - [ ] remove extraneous buttons: toggle spike (what does this do), show closest data on hover, compare data on hover

**Comment by hjwilli** [comment]
- [ ] display the number of unique values in or next to the 'type' column
- [ ] for numeric fields, if the number of unique values is small (<= 5), use stacked bar charts instead of box plots

**Comment by joshc0044** [comment]
I'm not 100% sure I understand how to take a stacked bar chart and display a pie chart. In the following example I'm using the `diabetes_small.csv` which I have posted on slack on aug 22 and will post here as well.

image

For this stacked chart there are 16 unique values in the dataset column for `payer_code`. If I understand the request correctly, the five most numerous values will be slices in the pie chart: `MC, ?, HM, BC, SP` and the rest will be combined into a single slice. Is it correct to take the 'stacks' of the bar chart, consolidate/combine them into one total, and display them as pie chart slices? And I will not incorporate the dependent column/target class info.

If this is not the case please clarify with this example what info from the stacked chart should be displayed in the pie chart

**Comment by joshc0044** [comment]
here is the small test dataset I used -
[diabetes_small.zip](https://github.com/EpistasisLab/pennai/files/3648469/diabetes_small.zip)
with the following dependent/target class: `readmitted`
with the following categorical features:
```
race,gender,age,weight,payer_code,medical_specialty,diag_1,diag_2,diag_3,max_glu_serum,A1Cresult,metformin,repaglinide,nateglinide,chlorpropamide,glimepiride,acetohexamide,glipizide,glyburide,tolbutamide,pioglitazone,rosiglitazone,acarbose,miglitol,troglitazone,tolazamide,examide,citoglipton,insulin,glyburide.metformin,glipizide.metformin,glimepiride.pioglitazone,metformin.rosiglitazone,metformin.pioglitazone,change,diabetesMed
```

**Comment by joshc0044** [comment]
as a quick first pass/rough draft here is a pie chart which should be similar to how it will look -
image

**Comment by joshc0044** [comment]
for the chart button menu 
>   * [ ]  remove extraneous buttons: toggle spike (what does this do), show closest data on hover, compare data on hover

the `compare data on hover` button looks like this for bar charts -
image
the `closest data on hover button` looks like this - 
image

for boxplots `compare data on hover` it looks like -
image

and for `closest data on hover button` boxplots - 
image

and `toggle spike lines` does this -
image

I'll disable the spike line button and the compare data buttons for the box plots; these buttons for this chart do not do much. For the stacked bar charts though, I can remove them or leave them as they have more actual utility

**Comment by hjwilli** [comment]
Archiving this pull request.

**Pull request closed without merging**